File: Maximum Ice Cream Bars/Maximum Ice Cream Bars.py
# Solution for Maximum Ice Cream Bars in PY

class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:

        freq = [0] * (max(costs) + 1)

        for cost in costs:
            freq[cost] += 1

        count = 0

        for price in range(1, len(freq)):

            can_buy = min(freq[price], coins // price)

            count += can_buy
            coins -= can_buy * price

            if coins < price:
                break

        return count

File: Maximum Ice Cream Bars/Maximum Ice Cream Bars.java
// Solution for Maximum Ice Cream Bars in JAVA

class Solution {
    public int maxIceCream(int[] costs, int coins) {

        int maxCost = 0;
        for (int cost : costs) maxCost = Math.max(maxCost, cost);

        int[] freq = new int[maxCost + 1];
        for (int cost : costs) freq[cost]++;

        int count = 0;

        for (int price = 1; price <= maxCost; price++) {

            int canBuy = Math.min(freq[price], coins / price);

            count += canBuy;
            coins -= canBuy * price;

            if (coins < price) break;
        }

        return count;
    }
}
